### ISSUE TITLE
[bug] The name "w3m-account-button" has already been used with this registry

### ISSUE THREAD

**Issue by mlenser** [WalletConnect/web3modal#1371]
### Link to minimal reproducible example

See https://github.com/WalletConnect/web3modal/issues/1371#issuecomment-1746525872

### Summary

```
Error: Failed to execute 'define' on 'CustomElementRegistry': the name "w3m-account-button" has already been used with this registry
```

It appears that the `@customElement('w3m-account-button')` code is being called twice via `createWeb3Modal`. I'm migrating from NextJS page router to app router and it appears that there may be some nesting going on where both are active at the same time which is causing `createWeb3Modal` to be run twice.

Can that custom element definition be gated by a check that it doesn't exist?

Thank you.

### List of related npm package versions

"@web3modal/wagmi": "^3.0.2"

**Comment by 0xAsimetriq** [comment]
Could you provide minimal reproducible example (in this case best would be a minimal repo with your setup)

**Comment by mlenser** [comment]
I am unable to reproduce the issue using the example apps which use vite. I tried calling `createWeb3Modal` twice and also tried it with one in a `useEffect`. Neither was able to trigger it.

It seems to be unique to a NextJS app which is half using Page router and half using App router.

I have setup https://github.com/mlenser/web3modal-w3m-account-button-already-used to reproduce this issue.

**Steps to reproduce:**
- Run `npm i` to install deps
- Run `dev`
- Go to http://localhost:3000/other in your browser
- See your terminal (Screenshot below)

<img width="873" alt="image" src="https://github.com/WalletConnect/web3modal/assets/235857/b4e7eb5f-206e-4b7b-bd4e-bb745985bd3e">

Text form:
```
'CustomElementRegistry' already has "w3m-account-button" defined. This may have been caused by live reload or hot module replacement in which case it can be safely ignored.
Make sure to test your application with a production build as repeat registrations will throw in production.
'CustomElementRegistry' already has "w3m-button" defined. This may have been caused by live reload or hot module replacement in which case it can be safely ignored.
Make sure to test your application with a production build as repeat registrations will throw in production.
'CustomElementRegistry' already has "w3m-connect-button" defined. This may have been caused by live reload or hot module replacement in which case it can be safely ignored.
Make sure to test your application with a production build as repeat registrations will throw in production.
'CustomElementRegistry' already has "wui-card" defined. This may have been caused by live reload or hot module replacement in which case it can be safely ignored.
Make sure to test your application with a production build as repeat registrations will throw in production.
'CustomElementRegistry' already has "wui-icon" defined. This may have been caused by live reload or hot module replacement in which case it can be safely ignored.
Make sure to test your application with a production build as repeat registrations will throw in production.
'CustomElementRegistry' already has "wui-image" defined. This may have been caused by live reload or hot module replacement in which case it can be safely ignored.
Make sure to test your application with a production build as repeat registrations will throw in production.
'CustomElementRegistry' already has "wui-loading-hexagon" defined. This may have been caused by live reload or hot module replacement in which case it can be safely ignored.
Make sure to test your application with a production build as repeat registrations will throw in production.
'CustomElementRegistry' already has "wui-loading-spinner" defined. This may have been caused by live reload or hot module replacement in which case it can be safely ignored.
Make sure to test your application with a production build as repeat registrations will throw in production.
'CustomElementRegistry' already has "wui-loading-thumbnail" defined. This may have been caused by live reload or hot module replacement in which case it can be safely ignored.
Make sure to test your application with a production build as repeat registrations will throw in production.
'CustomElementRegistry' already has "wui-shimmer" defined. This may have been caused by live reload or hot module replacement in which case it can be safely ignored.
Make sure to test your application with a production build as repeat registrations will throw in production.
'CustomElementRegistry' already has "wui-text" defined. This may have been caused by live reload or hot module replacement in which case it can be safely ignored.
Make sure to test your application with a production build as repeat registrations will throw in production.
'CustomElementRegistry' already has "wui-visual" defined. This may have been caused by live reload or hot module replacement in which case it can be safely ignored.
Make sure to test your application with a production build as repeat registrations will throw in production.
'CustomElementRegistry' already has "wui-flex" defined. This may have been caused by live reload or hot module replacement in which case it can be safely ignored.
Make sure to test your application with a production build as repeat registrations will throw in production.
'CustomElementRegistry' already has "wui-avatar" defined. This may have been caused by live reload or hot module replacement in which case it can be safely ignored.
Make sure to test your application with a production build as repeat registrations will throw in production.
'CustomElementRegistry' already has "wui-icon-box" defined. This may have been caused by live reload or hot module replacement in which case it can be safely ignored.
```

______________

_Note_: include a project id in `Providers.tsx` to avoid the app error of not having one. I cannot create a dummy one as wallet connect is in maintenance mode. When I inserted my real project id this app error went away, but the terminal errors remained.

<img width="484" alt="image" src="https://github.com/WalletConnect/web3modal/assets/235857/52d6bbd8-0101-4ae1-9328-137f502de26e">

**Comment by 0xAsimetriq** [comment]
`createWeb3Modal` keeps track of number of times it was initialized with, hence why in vite it works fine. I think your assumption about this being an app + pages based project (I can see 2 `<Provider />` components used within different places) is correct.

Likely nextjs somehow gets around this initialization check. Does it work correctly when you leave just single `<Provider />` in either app directory or normal one?

**Comment by mlenser** [comment]
Both app and pages need the `<Provider />`.

Without the `<Provider />` it shows
```
Error: Please call "createWeb3Modal" before using "useWeb3Modal" hook
```
<img width="777" alt="image" src="https://github.com/WalletConnect/web3modal/assets/235857/76a236a4-5940-43bd-82f4-8979c389410f">

This is true for `app` and `pages` - I tested removing `<Provider />` from both.

**Comment by mlenser** [comment]
This issue seems to also occur when fully using the `app` folder and no longer using the `pages` folder.

```
Lit is in dev mode. Not recommended for production! See https://lit.dev/msg/dev-mode for more information.
Multiple versions of Lit loaded. Loading multiple versions is not recommended. See https://lit.dev/msg/multiple-versions for more information.
 ⚠ Fast Refresh had to perform a full reload. Read more: https://nextjs.org/docs/messages/fast-refresh-reload
 ○ Compiling /lend/page ...
 ✓ Compiled /lend/page in 4.4s (6085 modules)
 ⨯ node_modules\@lit-labs\ssr-dom-shim\index.js (73:0) @ CustomElementRegistry.define
 ⨯ Error: Failed to execute 'define' on 'CustomElementRegistry': the name "w3m-account-button" has already been used with this registry
```

I'll look to add another reproducible case for this specific way of seeing the issue.

**Comment by 0xAsimetriq** [comment]
There is a template / starter project called "nexth" that uses web3modal in next's app directory, could you see if anything stands out in terms of why it could be working there but not in your setup? Could it be due to that useEffect workaround for wagmi?

This is how they implemented web3modal
https://github.com/wslyvh/nexth/blob/main/packages/app/src/context/Web3.tsx

**Comment by mlenser** [comment]
Ya, they're using `useEffect` and `useState` on the provider. However, that would force the whole app to only exist on the client. It's not a great practice.

The issue is the custom elements. Could a version of web3modal without the custom elements be exported? On React the hooks are the ideal way to interact with the modal. I'll try forking tomorrow and removing the custom elements to see if that would solve the issue in a better way.

**Comment by 0xAsimetriq** [comment]
Whole modal is build out of web-components to be compatible with any framework, custom element messages you see come from there (https://developer.mozilla.org/en-US/docs/Web/API/Web_components/Using_custom_elements)

I don't think wagmi works in server side environment (web3modal likely wont as well as it is a client side library relying on apis that might not be available in node). In addition to this, It seems like next's app directory is creating duplicates of these custom elements (they should only be created one time). It's unlikely that you'll be able to generate web3modal's code serverside and then propagate it to the client, it'll need to happen on client only (as is the case with many other frontend only libs relying on client apis)

**Comment by glitch-txs** [comment]
> that would force the whole app to only exist on the client.
Not true if you are using a layout.

**Comment by glitch-txs** [comment]
Hi @mlenser thanks for sharing the example that's super useful. 
I've played with it a bit, the main problem here is the live reload as stated in the warning

> "...This may have been caused by live reload..."

In your example you are running app routes together with the page routes which reloads the page when moving from one to another, this warn does not appear when using page routes alone.

![image](https://github.com/WalletConnect/web3modal/assets/66949816/ea11f108-006a-4293-8d29-2185ca7d3efb)

Have you experienced this warning on production?

**Comment by mlenser** [comment]
> I don't think wagmi works in server side environment

It doesn't, but using `useEffect` and `useState` to wrap the whole app means the whole app is client. None of it can be server rendered, even the pages that don't use wagmi. Additionally it makes the pages that do use wagmi worse for performance. It's not a real solution - it's a hack which will make stuff work, but in a poor way.

> Have you experienced this warning on production?

@glitch-txs 

Yes, this issue occurs in production. We see it on our Vercel preview environment with the modal not opening due to this issue. The pages build, but when actually used the web3modal does not open once this errors throws.

**Comment by glitch-txs** [comment]
> I will put together reproduction steps so it is clear.

dw I was able to reproduce it, we'll check internally how to remove this warn, thanks for your feedback 🤝

**Comment by mlenser** [comment]
@glitch-txs thanks for reproducing and for discussing it internally!

Can I ask what kind of timeline I might expect to address this? I only ask as we'll be in a bit of a limbo state as we've converted our whole site to app router, but can't deploy it due to this issue.

**Comment by mlenser** [comment]
I tested the https://github.com/wslyvh/nexth/blob/main/packages/app/src/context/Web3.tsx method, but the issue still exists.

Initially the modal still opens, but navigating between pages and sometimes refreshing causes the error to occur and the modal to no longer open.

I even tried putting `createWeb3Modal` in the `useEffect`. As soon as I navigate a few pages and then refresh my browser the error occurs and the modal breaks.

**Comment by fffedor** [comment]
Same problem. I just tried to migrate to web3modal v3 on next js 13+ and get the same error.  Haven't found a working solution yet.

**Comment by wslyvh** [comment]
Hi, creator of Nexth here. I experienced the same error yesterday, but haven't been able to reproduce properly. It does indeed seem like an intermittent issue during refresh/hot reload. Will continue to monitor...

Also providing some context on client/server architecture. Since most providers, contexts, auth, etc. will require client-side rendering you'll likely end up using some higher level client component either way. My understanding though is that you can still render server components within client components as long as you inject them as children and not import. I think this is a fair solution and definitely not a hacky work-around

See https://nextjs.org/docs/app/building-your-application/rendering/composition-patterns#supported-pattern-passing-server-components-to-client-components-as-props
Or https://app-router.vercel.app/context

**Comment by 0xAsimetriq** [comment]
Looking into this today/tomorrow 👍 ideally we find solution on our end that doesn't require any workarounds for users. Ty for extra context as well everyone 🤝

**Comment by 0xAsimetriq** [comment]
@mlenser I have just created this simple next repo with app router https://github.com/0xAsimetriq/web3modal-next-app-router-example and set up web3modal. TL;DR All I had to do is create this provider file and then use it in my apps layout https://github.com/0xAsimetriq/web3modal-next-app-router-example/blob/main/src/components/Provider.jsx

Everything works as expected, I tested hot reloading in connected and disconnected state etc. `use client` is a requirement in any file that needs `<WagmiConfig />` component.

Am I missing something here to why I'm not getting an issue you described? I think biggest diff is that your project has pages and app directories? In this case I'm not sure if there is any solution besides moving fully to one or the other maybe, not sure how next js treats this case under the hood when it comes to hot reloading.

**Comment by mlenser** [comment]
> Everything works as expected, I tested hot reloading in connected and disconnected state etc. use client is a requirement in any file that needs <WagmiConfig /> component.

My [reproduction repo](https://github.com/mlenser/web3modal-w3m-account-button-already-used) above does use `useClient`. `useClient` isn't an issue.

The steps to reproduce are similar to what I listed above. Slightly refined to be 100% reproducible:

**Steps to reproduce:**
- Run `npm i` to install deps
- Run `dev`
- Open http://localhost:3000 in your browser
- Click "Go to page router" (which goes to `/other`)
- See your terminal

```
Lit is in dev mode. Not recommended for production! See https://lit.dev/msg/dev-mode for more information.
Multiple versions of Lit loaded. Loading multiple versions is not recommended. See https://lit.dev/msg/multiple-versions for more information.
'CustomElementRegistry' already has "w3m-account-button" defined. This may have been caused by live reload or hot module replacement in which case it can be safely ignored.
Make sure to test your application with a production build as repeat registrations will throw in production.
'CustomElementRegistry' already has "w3m-button" defined. This may have been caused by live reload or hot module replacement in which case it can be safely ignored.
Make sure to test your application with a production build as repeat registrations will throw in production.
'CustomElementRegistry' already has "w3m-connect-button" defined. This may have been caused by live reload or hot module replacement in which case it can be safely ignored.
Make sure to test your application with a production build as repeat registrations will throw in production.
```

The dialog seems to open without issue, just the terminal warnings. The dialog also works on [the preview environment from that repo](https://web3modal-w3m-account-button-already-used.vercel.app). I created a branch with just app router (https://github.com/mlenser/web3modal-w3m-account-button-already-used/tree/only-app-router) and I can't get the warning to occur with the reproduction steps above nor when manually refreshing. So that specific warning seems to be isolated to using app router and page router.

Our actual app's preview environment does not open the dialog. We fully converted to App router to avoid any issues, but the dialog won't open. I've previously thought that was related to this warning, but am not able to confirm that currently.

I'll debug further and get back to you.

**Comment by 0xAsimetriq** [comment]
You shouldn't see this error in general (dev or prod)
> Multiple versions of Lit loaded. Loading multiple versions is not recommended. See https://lit.dev/msg/multiple-versions for more information.
'CustomElementRegistry' already has "w3m-account-button" defined.

Other ones related to lit being in dev mode are fine. So something is still off here.

Regarding modal not showing up in production, are there any errors you see in console / network tabs there? Are you env variables set up correctly for things like project id?

**Comment by mlenser** [comment]
The projectId is set - We use web3modal in production. The issue is somehow related to using it with App router, at least in our app.

The cause seems to be clicking a link or two and then we get:
```
 ⨯ node_modules\@lit-labs\ssr-dom-shim\index.js (73:0) @ CustomElementRegistry.define
 ⨯ Error: Failed to execute 'define' on 'CustomElementRegistry': the name "w3m-account-button" has already been used with this registry
```
and then the modal stop opening.

**Comment by 0xAsimetriq** [comment]
@mlenser Do you by any chance have a link to your prod deployment I could look at?

**Comment by mlenser** [comment]
The error only shows on the terminal in `dev`. On a build, and on the preview, there is no console error. The dialog works until something happens which I believe is that error being thrown.

I'll try to setup a repo with the error.

**Comment by mlenser** [comment]
I believe "Multiple versions of Lit loaded" is caused by using `@tanstack/react-query` alongside `web3modal`.

I've been able to avoid that message by resolving the `lit` versions in `package.json`:
```
  "resolutions": {
    "@lit-labs/ssr-dom-shim": "^1.1.0",
    "lit-html": "^2.8.0",
    "@tanstack/react-query": "^4.35.7"
  }
```

You can see the issue on https://github.com/mlenser/web3modal-w3m-account-button-already-used/tree/with-react-query-and-resolutions if you remove the `resolutions` in `package.json`.

_______

That doesn't solve the dialog not opening issue though. That seems to be caused by some divs, icons, and other elements inside the connect button. If I only have text such as `Connect` inside of it and restart my server the issue appears to go away.

I'll see if I can create a reproduction.

**Comment by mlenser** [comment]
Summary of my findings related to `web3modal`:

- There is definitely at least a warning in `web3modal` about "The name "w3m-account-button" has already been used with this registry" if you're running app router and pages router. This may pop up in other circumstances (app router), but it was intermittent and the exposure of it may have been related to other issues below.
- There are possible mismatches in `lit` dependencies when running `web3modal` and `@tanstack/react-query` which lead to "Multiple versions of Lit loaded" which can cause bugs. Check your `yarn.lock` and `package-lock.json` for multiple versions of `lit`. I found multiple versions of `@lit-labs/ssr-dom-shim` even on the latest of each.
- If the contents of a connect button change when the user is connected then the contents should be wrapped in a `useEffect` and `useState`. See https://nextjs.org/docs/messages/react-hydration-error#solution-1-using-useeffect-to-run-on-the-client-only.
- This hydration issue does not cause any of our other libs to break. We have a few on some pages without issues in any of our other libs. I'm not sure why hydration issues would cause issues uniquely for `web3modal`.

**Comment by bartomolina** [comment]
I'm getting the same errors in the console when using WC alongside next-pwa (https://github.com/DuCanhGH/next-pwa). This is my sample repo: https://github.com/bartomolina/wc-pwa-issue

- run `pnpm install`
- set WC projectid in .env
- run `pnpm dev`
- click on setting and then click on log out to navigate back to the homepage

**Comment by 0xAsimetriq** [comment]
@bartomolina ty for the repro, I'm able to see the errors now. Are you seeing any issues when deploying this to prod i.e. I'm trying to figure out if `This may have been caused by live reload or hot module replacement in which case it can be safely ignored.` is valid here or not.

**Comment by bartomolina** [comment]
Haven't deployed in prod yet, but the app is working fine otherwise, so this is not a showstopper for me.

**Comment by 0xAsimetriq** [comment]
Alright, I believe this is now solved (part of next release, if you want to test canary use `3.1.0-598ec14`). Had to re-create lit's `createElement` decorator and add a check for existing custom components. Tested with @bartomolina example on that canary, no errors 👍

**Comment by 0xAsimetriq** [comment]
@mlenser as for your issues on prod, it's likely something else, could definitely be conflict between lit versions, but as far as I can tell `@tanstack/react-query` doesn't seem to have dependency on it.

**Comment by mlenser** [comment]
Thanks for the investigation and change to `createElement`.

My prod issue was somehow connected to a hydration mismatch & the duplicate create element. I've got it solved, but the release to prevent the warnings/error would possibly solve it as well.

Thanks!
